aptos-move/aptos-release-builder/src/simulate.rs

```diff
@@ -0,0 +1,500 @@
+        Some(idx) => idx,
+        None => {
+            let idx = pool.len();
+            pool.push(val);
+            idx
+        },
+    }
+}
+
+fn get_or_add_addr(m: &mut CompiledModule, addr: AccountAddress) -> AddressIdentifierIndex {
+    AddressIdentifierIndex::new(get_or_add(&mut m.address_identifiers, addr) as u16)
+}
+
+fn get_or_add_ident(m: &mut CompiledModule, ident: Identifier) -> IdentifierIndex {
+    IdentifierIndex::new(get_or_add(&mut m.identifiers, ident) as u16)
+}
+
+fn get_or_add_module_handle(
+    m: &mut CompiledModule,
+    addr: AccountAddress,
+    name: Identifier,
+) -> ModuleHandleIndex {
+    let addr = get_or_add_addr(m, addr);
+    let name = get_or_add_ident(m, name);
+    let module_handle = ModuleHandle {
+        address: addr,
+        name,
+    };
+    ModuleHandleIndex::new(get_or_add(&mut m.module_handles, module_handle) as u16)
+}
+
+fn get_or_add_signature(m: &mut CompiledModule, sig: Vec<SignatureToken>) -> SignatureIndex {
+    SignatureIndex::new(get_or_add(&mut m.signatures, Signature(sig)) as u16)
+}
+
+fn get_or_add_simple_function_handle(
+    m: &mut CompiledModule,
+    module_addr: AccountAddress,
+    module_name: Identifier,
+    func_name: Identifier,
+    params: Vec<SignatureToken>,
+    returns: Vec<SignatureToken>,
+) -> FunctionHandleIndex {
+    let module = get_or_add_module_handle(m, module_addr, module_name);
+    let name = get_or_add_ident(m, func_name);
+    let parameters = get_or_add_signature(m, params);
+    let return_ = get_or_add_signature(m, returns);
+    let func_handle = FunctionHandle {
+        module,
+        name,
+        parameters,
+        return_,
+        type_parameters: vec![],
+        access_specifiers: None,
+    };
+    FunctionHandleIndex::new(get_or_add(&mut m.function_handles, func_handle) as u16)
+}
+
+fn list_scripts(proposal_path: &Path) -> Result<Vec<PathBuf>> {
+    let mut script_paths = std::fs::read_dir(proposal_path)?
+        .filter_map(|entry| entry.ok())
+        .filter_map(|entry| {
+            let path = entry.path();
+            if path.extension().map(|s| s == "move").unwrap_or(false) {
+                Some(path)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    script_paths.sort();
+
+    Ok(script_paths)
+}
+/***************************************************************************************************
+ * Simulation State View
+ *
+ **************************************************************************************************/
+/// A state view specifically designed for managing the side effects generated by
+///  the governance scripts.
+///
+/// It comprises two components:
+/// - A remote debugger state view to enable on-demand data fetching.
+/// - A local state store to allow new changes to be stacked on top of the remote state.
+struct SimulationStateView<'a, S> {
+    remote: &'a S,
+    states: Mutex<HashMap<StateKey, Option<StateValue>>>,
```

Comment:
nit: maybe use `parking_lot::Mutex` to avoid unwraps on poisoning and just release the lock?